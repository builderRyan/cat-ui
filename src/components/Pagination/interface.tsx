import * as React from 'react'

export interface StyledPageItem {
  active?: boolean
}
