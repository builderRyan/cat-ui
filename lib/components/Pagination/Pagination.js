"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var react_1 = require("react");
var classnames_1 = require("classnames");
// import * as memoize from 'memoize-one'
var memoize_one_1 = require("memoize-one");
var isNumber = require("lodash/isNumber");
var styled_1 = require("./styled");
var PageItem_1 = require("./PageItem");
/**
 * 分页
 */
var Pagination = /** @class */ (function (_super) {
    __extends(Pagination, _super);
    function Pagination(props) {
        var _this = _super.call(this, props) || this;
        _this.onItemClick = function (value) {
            var onChange = _this.props.onChange;
            onChange && onChange(value);
        };
        _this.onPrevClick = function () {
            var _a = _this.props, onChange = _a.onChange, pageSize = _a.pageSize, current = _a.current;
            // 不是第一个
            if (current > 1) {
                onChange && onChange(current - 1, pageSize);
            }
        };
        _this.onNextClick = function () {
            var _a = _this.props, onChange = _a.onChange, pageSize = _a.pageSize, current = _a.current, total = _a.total;
            var pageNum = Math.ceil(total / pageSize);
            // 不是最后一个
            if (current < pageNum) {
                onChange && onChange(current + 1, pageSize);
            }
        };
        // 计算带省略号的页码
        _this.calculateShowPages = function (current, pageNum, bufferSize) {
            var res = [current];
            for (var i = 1; i <= bufferSize; i++) {
                if (current - i > 1) {
                    res = [current - i].concat(res);
                }
                if (current + i < pageNum) {
                    res = res.concat([(current + i)]);
                }
            }
            if (current - (bufferSize + 1) > 1) {
                res = ['...'].concat(res);
            }
            if (current > 1) {
                res = [1].concat(res);
            }
            if (current + bufferSize + 1 < pageNum) {
                res = res.concat([' ...']);
            }
            if (current < pageNum) {
                res = res.concat([pageNum]);
            }
            return res;
        };
        _this.renderItems = memoize_one_1.default(function (current, total, pageSize) {
            var pageNum = Math.ceil(total / pageSize);
            var res = _this.calculateShowPages(current, pageNum, 2);
            return res.map(function (element, index) {
                if (isNumber(element)) {
                    return (React.createElement(PageItem_1.default, { key: element, value: element, active: current === element, onItemClick: _this.onItemClick }, element));
                }
                else {
                    return (React.createElement(styled_1.StyledIcon, { key: "more+" + index, type: 'more' }));
                }
            });
        });
        _this.state = {};
        return _this;
    }
    Pagination.prototype.render = function () {
        var _a = this.props, className = _a.className, style = _a.style, current = _a.current, total = _a.total, pageSize = _a.pageSize;
        var classes = classnames_1.default('hmly-pagination', className);
        var items = this.renderItems(current, total, pageSize);
        // console.log('current, total, pageSize', memoizeOne)
        return (React.createElement(styled_1.StyledPagination, { className: classes, style: style },
            React.createElement(styled_1.StyledIcon, { type: 'prev', onClick: this.onPrevClick }),
            items,
            React.createElement(styled_1.StyledIcon, { type: 'next', onClick: this.onNextClick })));
    };
    Pagination.defaultProps = {
        pageSize: 10
    };
    return Pagination;
}(react_1.Component));
exports.default = Pagination;
//# sourceMappingURL=Pagination.js.map