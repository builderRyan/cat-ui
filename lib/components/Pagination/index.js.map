{"version":3,"file":"index.js","sourceRoot":"","sources":["../../../src/components/Pagination/index.ts"],"names":[],"mappings":";;AAAA,2CAAqC;AAErC,kBAAe,oBAAU,CAAA"}