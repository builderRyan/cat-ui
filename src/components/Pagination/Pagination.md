Example

```jsx
class Example extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      current1: 2
    }
    this.onChange1 = this.onChange1.bind(this)
  }

  onChange1 (page) {
    this.setState({ current1 : page})
    console.log('Example:onChange1', page)
  }

  render () {
    const { current1 } = this.state
    return (<div className='pagination-box'>
      <Pagination
        current={current1}
        total={85}
        onChange={this.onChange1} />

      <style>
        {`
          .pagination-box {
            padding: 5px;
          }
        `}
      </style>
    </div>
    )
  }
}

<Example />
```
