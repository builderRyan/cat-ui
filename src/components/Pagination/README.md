# Pagination 分页

## Example

```jsx
<Pagination
  current={current}
  total={total}
  onChange={(page, pageSize) => setCurrent(page)}
/>
```

## API

|   属性    |                    类型                    | 默认值 |                       描述                       | 是否必填 |
| :-------: | :----------------------------------------: | :----: | :----------------------------------------------: | :------: |
|  prefix   |                  `string`                  | `cat`  |                     类名前缀                     |    -     |
| className |                  `string`                  |   -    |                       类名                       |    -     |
|   style   |           `React.CSSProperties`            |   -    |                     行内样式                     |    -     |
|  current  |                  `number`                  |  `1`   |                     当前页码                     |    -     |
|   total   |                  `number`                  |   -    |                     当前页码                     |    是    |
| pageSize  |                  `number`                  |  `10`  |                     每页条数                     |    -     |
| onChange  | `(page: number, pageSize: number) => void` |   -    | 页码改变的回调。参数是改变后的页码以及每页的条数 |    -     |
