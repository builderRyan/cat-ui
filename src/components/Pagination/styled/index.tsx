import styled from 'styled-components'
import { primaryBtnColor } from '../../../styles/theme'
import { StyledPageItem } from '../interface'
import Icon from '../../Icon'

const StyledPagination = styled.div``;

const StyledPageItem = styled.div`
  display: inline-block;
  font-size: 14px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: normal;
  color: ${(props: StyledPageItem) => props.active ? '#313646' : '#bcbcbc' };
  padding: 0 10px;
  cursor: pointer;
`

const StyledIcon = styled(Icon)`
  cursor: pointer;
  padding: 0 10px;
`

export { StyledPagination, StyledPageItem, StyledIcon }
