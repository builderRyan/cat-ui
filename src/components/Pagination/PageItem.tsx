import * as React from 'react'
import { Component } from 'react'
import classNames from 'classnames'
import { StyledPageItem } from './styled'

export interface PageItemProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** active */
  active?: boolean;
  /** value */
  value?: number;
  /** onItemClick */
  onItemClick?: any;
}

/**
 * PageItem
 */
class PageItem extends Component<PageItemProps, any> {
  handleClick = () => {
    const { onItemClick, value } = this.props
    onItemClick && onItemClick(value)
  }

  render() {
    const { className, style, active, children } = this.props
    const classes = classNames('hmly-page-item', className)

    return (
      <StyledPageItem
        className={classes}
        style={style}
        active={active}
        onClick={this.handleClick}>
        {children}
      </StyledPageItem>
    )
  }
}

export default PageItem
