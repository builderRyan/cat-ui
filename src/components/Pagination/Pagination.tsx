import * as React from 'react';
import { Component } from 'react';
import classNames from 'classnames';
import memoizeOne from 'memoize-one';
import * as isNumber from 'lodash/isNumber';
import PageItem from './PageItem';
import Icon from '../Icon';

export interface PaginationProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** 当前页数 */
  current?: number;
  /** 数据总数 */
  total?: number;
  /** 每页条数 */
  pageSize?: number;
  /** 分页大小改变时候的回调函数 */
  onPageSizeChange?: (current: number) => void;
  /** 每页条数 */
  onChange?: (page: number, pageSize?: number) => void;
}

/**
 * 分页
 */
class Pagination extends Component<PaginationProps, any> {
  static defaultProps = {
    pageSize: 10
  };

  constructor(props: PaginationProps) {
    super(props);
    this.state = {};
  }

  onItemClick = (value: number) => {
    const { onChange } = this.props;
    onChange && onChange(value);
  };

  onPrevClick = () => {
    const { onChange, pageSize, current } = this.props;
    // 不是第一个
    if (current > 1) {
      onChange && onChange(current - 1, pageSize);
    }
  };

  onNextClick = () => {
    const { onChange, pageSize, current, total } = this.props;
    const pageNum = Math.ceil(total / pageSize);
    // 不是最后一个
    if (current < pageNum) {
      onChange && onChange(current + 1, pageSize);
    }
  };

  // 计算带省略号的页码
  calculateShowPages = (
    current: number,
    pageNum: number,
    bufferSize: number
  ) => {
    let res: Array<number | string> = [current];
    for (let i = 1; i <= bufferSize; i++) {
      if (current - i > 1) {
        res = [current - i, ...res];
      }
      if (current + i < pageNum) {
        res = [...res, current + i];
      }
    }
    if (current - (bufferSize + 1) > 1) {
      res = ['...', ...res];
    }
    if (current > 1) {
      res = [1, ...res];
    }
    if (current + bufferSize + 1 < pageNum) {
      res = [...res, ' ...'];
    }
    if (current < pageNum) {
      res = [...res, pageNum];
    }
    return res;
  };

  renderItems = memoizeOne(
    (current: number, total: number, pageSize: number) => {
      const pageNum = Math.ceil(total / pageSize);
      const res = this.calculateShowPages(current, pageNum, 2);
      return res.map((element, index) => {
        if (isNumber(element)) {
          return (
            <PageItem
              key={element}
              value={element}
              active={current === element}
              onItemClick={this.onItemClick}
            >
              {element}
            </PageItem>
          );
        } else {
          return <Icon key={`more+${index}`} type="more" />;
        }
      });
    }
  );

  render() {
    const { className, style, current, total, pageSize } = this.props;
    const classes = classNames('cat-pagination', className);
    const items = this.renderItems(current, total, pageSize);

    return (
      <div className={classes} style={style}>
        <Icon type="prev" onClick={this.onPrevClick} />
        {items}
        <Icon type="next" onClick={this.onNextClick} />
      </div>
    );
  }
}

export default Pagination;
