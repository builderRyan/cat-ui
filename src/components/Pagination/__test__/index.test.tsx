import * as React from 'react'
import * as Enzyme from 'enzyme'
import { render, mount } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import Pagination from '../index'

Enzyme.configure({ adapter: new Adapter() })

describe('Pagination', () => {
  // test('renders correctly', () => {
  //   const wrapper = render(
  //     <Pagination
  //       current={2}
  //       total={85} />
  //   )
  //   expect(wrapper).toMatchSnapshot()
  // })

  test('two plus two is four', () => {
    expect(2 + 2).toBe(4)
  })
})
