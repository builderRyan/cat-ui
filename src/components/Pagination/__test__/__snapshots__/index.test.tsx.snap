// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<Pagination
  current={1}
  pageSize={10}
  prefix="cat"
  total={100}
>
  <div
    className="cat-pagination"
  >
    <Icon
      className="cat-pagination__prev cat-pagination__prev--disabled"
      onClick={[Function]}
      type="prev"
    >
      <i
        className="cat-icon cat-icon-left cat-pagination__prev cat-pagination__prev--disabled"
        onClick={[Function]}
      />
    </Icon>
    <PageItem
      active={true}
      key="1"
      onClick={[Function]}
      prefix="cat-pagination"
      value={1}
    >
      <span
        className="cat-pagination__item cat-pagination__item--active"
        onClick={[Function]}
      >
        1
      </span>
    </PageItem>
    <PageItem
      active={false}
      key="2"
      onClick={[Function]}
      prefix="cat-pagination"
      value={2}
    >
      <span
        className="cat-pagination__item"
        onClick={[Function]}
      >
        2
      </span>
    </PageItem>
    <PageItem
      active={false}
      key="3"
      onClick={[Function]}
      prefix="cat-pagination"
      value={3}
    >
      <span
        className="cat-pagination__item"
        onClick={[Function]}
      >
        3
      </span>
    </PageItem>
    <span
      className="cat-pagination__item cat-pagination__item--no-hover"
      key="omit3"
    >
      …
    </span>
    <PageItem
      active={false}
      key="10"
      onClick={[Function]}
      prefix="cat-pagination"
      value={10}
    >
      <span
        className="cat-pagination__item"
        onClick={[Function]}
      >
        10
      </span>
    </PageItem>
    <Icon
      className="cat-pagination__next"
      onClick={[Function]}
      type="next"
    >
      <i
        className="cat-icon cat-icon-right cat-pagination__next"
        onClick={[Function]}
      />
    </Icon>
  </div>
</Pagination>
`;

exports[`Pagination renders correctly 2`] = `
<Pagination
  current={1}
  pageSize={10}
  prefix="cat"
  total={100}
>
  <div
    className="cat-pagination"
  >
    <Icon
      className="cat-pagination__prev cat-pagination__prev--disabled"
      onClick={[Function]}
      type="prev"
    >
      <i
        className="cat-icon cat-icon-left cat-pagination__prev cat-pagination__prev--disabled"
        onClick={[Function]}
      />
    </Icon>
    <PageItem
      active={true}
      key="1"
      onClick={[Function]}
      prefix="cat-pagination"
      value={1}
    >
      <span
        className="cat-pagination__item cat-pagination__item--active"
        onClick={[Function]}
      >
        1
      </span>
    </PageItem>
    <PageItem
      active={false}
      key="2"
      onClick={[Function]}
      prefix="cat-pagination"
      value={2}
    >
      <span
        className="cat-pagination__item"
        onClick={[Function]}
      >
        2
      </span>
    </PageItem>
    <PageItem
      active={false}
      key="3"
      onClick={[Function]}
      prefix="cat-pagination"
      value={3}
    >
      <span
        className="cat-pagination__item"
        onClick={[Function]}
      >
        3
      </span>
    </PageItem>
    <span
      className="cat-pagination__item cat-pagination__item--no-hover"
      key="omit3"
    >
      …
    </span>
    <PageItem
      active={false}
      key="10"
      onClick={[Function]}
      prefix="cat-pagination"
      value={10}
    >
      <span
        className="cat-pagination__item"
        onClick={[Function]}
      >
        10
      </span>
    </PageItem>
    <Icon
      className="cat-pagination__next"
      onClick={[Function]}
      type="next"
    >
      <i
        className="cat-icon cat-icon-right cat-pagination__next"
        onClick={[Function]}
      />
    </Icon>
  </div>
</Pagination>
`;
